File: python/createTrie.py
# This question is asked by Microsoft.
# Implement a trie class that supports insertion and search functionalities.
# Note: You may assume only lowercase alphabetical characters will added to your trie.

# Ex: Given the following operations on your trie…

# Trie trie = new Trie()
# trie.insert("programming");
# trie.search("computer") // returns false.
# trie.search("programming") // returns true.


class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_word = True

    def search(self, word):
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_word

    def starts_with(self, prefix):
        node = self.root
        for char in prefix:
            if char not in node.children:
                return False
            node = node.children[char]
        return True


trie = Trie()
trie.insert("programming")  # will insert "programming"
print(trie.search("programming"))  # True
print(trie.search("computer"))  # False
print(trie.starts_with("program"))  # True

File: python/createTrie.ts
// # This question is asked by Microsoft.
// # Implement a trie class that supports insertion and search functionalities.
// # Note: You may assume only lowercase alphabetical characters will added to your trie.

// # Ex: Given the following operations on your trie…

// # Trie trie = new Trie()
// # trie.insert("programming");
// # trie.search("computer") // returns false.
// # trie.search("programming") // returns true.

class TrieNode {
  constructor(public children: object, public isWord: boolean) {
    this.children = {};
    this.isWord = false;
  }
}

class Trie {
  root: TrieNode;
  constructor() {
    this.root = new TrieNode({}, false);
  }

  public insert(word: string) {
    let currentNode = this.root;
    for (let i = 0; i < word.length; i++) {
      const char = word[i];
      if (!currentNode.children.hasOwnProperty(char)) {
        currentNode.children[char] = new TrieNode({}, false);
      }
      currentNode = currentNode.children[char];
    }

    currentNode.isWord = true;
  }

  public search(word: string) {
    let currentNode = this.root;
    for (let i = 0; i < word.length; i++) {
      const char = word[i];
      if (!currentNode.children[char]) {
        return false;
      }
      currentNode = currentNode.children[char];
    }
    return currentNode.isWord;
  }

  public startsWith(prefix: string) {
    let currentNode = this.root;
    for (let i = 0; i < prefix.length; i++) {
      const char = prefix[i];
      if (!currentNode.children[char]) {
        return false;
      }
      currentNode = currentNode.children[char];
    }
    return true;
  }

  public print() {
    console.log(this.root);
  }
}

const trie = new Trie();
trie.insert("programming");
trie.insert("program");
trie.insert("programmer");

trie.search("programming"); // returns true.
trie.search("program"); // returns true.
trie.search("programmer"); // returns true.
trie.search("programmingx"); // returns false.

trie.startsWith("pra"); // returns false.
trie.startsWith("pro"); // returns true.
trie.startsWith("prog"); // returns true.
trie.startsWith("progra"); // returns false.

trie.print();

// const trie = new Trie();
// trie.insert("programming");
// trie.insert("program");
// trie.insert("programmer");
// console.log(trie);
// console.log(trie.search("programming"));
// console.log(trie.search("program"));
// console.log(trie.search("programmer"));
// console.log(trie.startsWith("program"));
// console.log(trie.startsWith("programming"));
// console.log(trie.startsWith("programmer"));
// console.log(trie.startsWith("programminger"));
